### PR TITLE
Add tcp support

### LINKED ISSUE #52: TCP support
Hello

Would you be interested in TCP support?
I have some sample code which implements a TCP client

TCP would allow to lift the restriction on the maximum object size that can be sent through OSC

### PR DESCRIPTION
**Pull request by bobh66**
Added support for TCP client and server, using both the 1.0 "length" protocol and the (default) 1.1 SLIP protocol.

Blocking/threaded and asyncio implementation are available for client and server.

Examples and tests are included.  Basic functional testing of the client was done using QLab 5.4.4 and Eos 3.2.7, and I verified that it is possible to trigger cues in both using a TCP client. 

Fixes #52

### CODE REVIEW

**Comment by bobh66** [comment]
I was also thinking about updating the UDP server to allow for responses to be sent the same way I have it working for TCP, for consistency.  That could also be a separate PR.

File: pythonosc/osc_tcp_server.py [hunk 1]
```diff
@@ -0,0 +1,392 @@
+# TODO: timeouts!
```
Comment (attwad) [inline pythonosc/osc_tcp_server.py:1]: please remove the comment and leave an open issue for anyone willing to take it instead, thank you :)

File: pythonosc/osc_tcp_server.py [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+            raise ValueError("OSC Mode must be '1.0' or '1.1'")
+        if self.mode == MODE_1_0:
+            super().__init__(server_address, _TCPHandler1_0)
+        else:
+            super().__init__(server_address, _TCPHandler1_1)
+        self._dispatcher = dispatcher
+
+    def verify_request(self, request, client_address):
+        """Returns true if the data looks like a valid OSC TCP datagram."""
+        # d = request.recv(9999).decode("utf-8")
```
Comment (attwad) [inline pythonosc/osc_tcp_server.py:193]: remove commented out lines/fix method description or remove method altogether if it's unused?

File: pythonosc/osc_tcp_server.py [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+        :param server_address: tuple of (IP address to bind to, port)
+        :param dispatcher: a pythonosc.dispatcher.Dispatcher
+        """
+        self._port = port
+        self._server_address = server_address
+        self._dispatcher = dispatcher
+        self._server = None
+        self._mode = mode
+
+    # class _OSCProtocolFactory(asyncio.DatagramProtocol):
```
Comment (attwad) [inline pythonosc/osc_tcp_server.py:247]: remove commented out lines please

File: pythonosc/slip.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
+#  Copyright (c) 2020. Ruud de Jong
```
Comment (attwad) [inline pythonosc/slip.py:1]: was this copied verbatim? that github project seems to be MIT so the MIT license should appear somewhere where the copy ended up, please copy it here if so.

**Review by attwad — CHANGES_REQUESTED** [review]

**Comment by attwad** [comment]
The mypy integration is starting to get a bit annoying for contributors I fear... If it's too much work to get it fixed please let me know and I'll look into how to simplify/remove this if needed.
Thanks for the PR btw!

File: pythonosc/osc_tcp_server.py (hunk 1 above)
Comment (bobh66) [reply pythonosc/osc_tcp_server.py]: This was an artifact from the earlier implementation - I don't know exactly what it means to I removed it.

File: pythonosc/osc_tcp_server.py (hunk 3 above)
Comment (bobh66) [reply pythonosc/osc_tcp_server.py]: Done

File: pythonosc/slip.py (hunk 4 above)
Comment (bobh66) [reply pythonosc/slip.py]: Added the license info

File: pythonosc/osc_tcp_server.py (hunk 2 above)
Comment (bobh66) [reply pythonosc/osc_tcp_server.py]: Removed unused method

**Comment by attwad** [comment]
Thanks! I suggest syncing with the main branch because I formatted everything with black to ensure consistency amongst PRs and across the codebase, just 3.10 seems to be failing with a `timeout` module non existent, please double check

**Comment by cococow123** [comment]
This is great! Thanks for making a PR for the TCP version!
Since the TCP client can send and receive OSC packets over the same address/port, is it possible to have the client also support a dispatcher to handle incoming messages?

**Comment by bobh66** [comment]
> This is great! Thanks for making a PR for the TCP version! Since the TCP client can send and receive OSC packets over the same address/port, is it possible to have the client also support a dispatcher to handle incoming messages?

I think that's possible but I would implement it as another class that combines one of the existing TCP Clients with a Dispatcher, so that the clients can still be used separately.  That's probably a good candidate for a separate PR, this one has already grown more than I had expected.

**Comment by bobh66** [comment]
@attwad I still need to add the CHANGELOG entry and I realized I should add receive support to the UDP Client since I added the response support to the UDP Server.

**Review by attwad — COMMENTED** [review]
Code lgtm thanks for all the additions, I recommend using https://github.com/nektos/act to run the test workflows locally and make sure they all pass, as first time contributors require my approval to run the github actions it takes time everytime, that should speed up the review process :) Thanks!

File: pythonosc/osc_tcp_server.py [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+
+from pythonosc import osc_message_builder, slip
+from pythonosc.dispatcher import Dispatcher
+
+LOG = logging.getLogger()
+MODE_1_0 = "1.0"
+MODE_1_1 = "1.1"
+
+
+# def _call_handlers_for_packet(data: bytes, dispatcher: Dispatcher, client_address: Tuple[str, int]) -> List:
```
Comment (attwad) [inline pythonosc/osc_tcp_server.py:50]: please remove unused code

**Comment by attwad** [comment]
Thanks!

**Comment by attwad** [comment]
I just pushed this as part of the 1.9.0 release

**Pull request merged**